beets/ui/__init__.py

```diff
@@ -465,91 +478,192 @@ def human_seconds_short(interval):
+                    # Legacy color definition (type: unicode). Convert.
+                    if color_def in LEGACY_COLORS:
+                        color_def = LEGACY_COLORS[color_def]
+                    else:
+                        raise ValueError("no such color %s", color_def)
+
+                COLORS[name] = color_def
+        # In case a 3rd party plugin is still passing the actual color ('red')
+        # instead of the abstract color name ('text_error')
+        color = COLORS.get(color_name)
+        if not color:
+            log.debug(u"Invalid color_name: {0}", color_name)
+            color = color_name
+        return _colorize(color, text)
+    else:
         return text
 
-    global COLORS
-    if not COLORS:
-        COLORS = {name:
-                  config['ui']['colors'][name].as_str()
-                  for name in COLOR_NAMES}
-    # In case a 3rd party plugin is still passing the actual color ('red')
-    # instead of the abstract color name ('text_error')
-    color = COLORS.get(color_name)
-    if not color:
-        log.debug('Invalid color_name: {0}', color_name)
-        color = color_name
-    return _colorize(color, text)
-
-
-def _colordiff(a, b, highlight='text_highlight',
-               minor_highlight='text_highlight_minor'):
+
+def uncolorize(colored_text):
+    """Remove colors from a string."""
+    # Define a regular expression to match ANSI codes.
+    # See: http://stackoverflow.com/a/2187024/1382707
+    # Explanation of regular expression:
+    #     \x1b     - matches ESC character
+    #     \[       - matches opening square bracket
+    #     [;\d]*   - matches a sequence consisting of one or more digits or
+    #                semicola
+    #     [A-Za-z] - matches a letter
+    ansi_code_regex = re.compile(r"\x1b\[[;\d]*[A-Za-z]", re.VERBOSE)
+    # Strip ANSI codes from `colored_text` using the regular expression.
+    text = ansi_code_regex.sub(u"", colored_text)
+    return text
+
+
+def color_len(colored_text):
+    """Measure the length of a string while excluding ANSI codes from the
+    measurement. The standard `len(my_string)` method also counts ANSI codes
+    to the string length, which is counterproductive when layouting a
+    Terminal interface.
+    """
+    # Return the length of the uncolored string.
+    return len(uncolorize(colored_text))
+
+
+def _colordiff(a, b):
     """Given two values, return the same pair of strings except with
     their differences highlighted in the specified color. Strings are
     highlighted intelligently to show differences; other values are
     stringified and highlighted in their entirety.
     """
-    if not isinstance(a, str) \
-       or not isinstance(b, str):
+    # Set highlight colors.
+    highlight_added = "text_diff_added"
+    highlight_removed = "text_diff_removed"
+    minor_highlight = "text_highlight_minor"
+
+    if not isinstance(a, str) or not isinstance(b, str):
         # Non-strings: use ordinary equality.
```

Comment:
This is what the code did before, but code and comment are in conflict (since the stringified objects are compared). Maybe one should check the `git blame` to figure out what the original intention was here? I would guess that the right thing to do is what the comment says, but still return the stringified (and possibly highlighted) versions.

Reply:
Git blame did indeed show that two different people added the comment and functionality. The original intention seems to be that e.g. ints are compared as-is. I've moved the bytes -> readable paths upwards, and then kept the catch-all "not strings" comparison to recreate the original functionality.